system/security/plugins/testauthSecurity/testauthSecurity.cpp

```diff
@@ -0,0 +1,591 @@
+        while (*p)
+        {
+            if (*p != ':')
+            {
+                scope.append(*(p++));
+                continue;
+            }
+
+            if (*(p+1) != ':')
+                return false;
+            if (fullPathOnly)
+                scope.append('/');
+            else
+            {
+                scopes.append(scope);
+                scope.clear();
+            }
+            p += 2;
+        }
+        if (fullPathOnly && (scope.charAt(scope.length() - 1) == '/'))
+            return false;
+        if (!fullPathOnly && scope.isEmpty())
+            return false;
+        scopes.append(scope);
+        return true;
+    }
+
+    SecAccessFlags getFileScopeAccess(IPropertyTree* accessTree, const StringArray& nodes, unsigned pos, SecAccessFlags access, bool first)
+    {
+        if (!nodes.isItem(pos))
+            return access;
+        const char* node = nodes.item(pos);
+        if (isEmptyString(node))
+            return access;
+        IPropertyTree* element = accessTree->queryPropTree(node);
+        if (!element)
+            return access;
+        if (element->hasProp("@access"))
+        {
+            int newAccess = element->getPropInt("@access");
+            if (newAccess < SecAccess_Read)
+                return (SecAccessFlags) newAccess;
+            if (first || newAccess < access) //least access wins
+            {
+                access = (SecAccessFlags) newAccess;
+                first = false;
+            }
+        }
+        return getFileScopeAccess(element, nodes, ++pos, access, first);
+    }
+
+    SecAccessFlags queryECLWUScopeAccess(const char* resource)
+    {
+        int* allow = eclwuScopeAccesses.getValue(resource);
+        return allow ? (SecAccessFlags) *allow : defaulteECLWUScopeAccess;
+    }
+};
+
+typedef CUserAccess* CUserAccessPtr;
+
+class CTestAuthSecurityManager : public CBaseSecurityManager
+{
+    static const SecFeatureSet s_implementedFeatureMask = SMF_QuerySecMgrType | SMF_QuerySecMgrTypeName | SMF_CreateAuthMap |
+                                                          SMF_CreateFeatureMap | SMF_LogoutUser | SMF_GetDescription | SMF_Authorize |
+                                                          SMF_AuthorizeEx_Named | SMF_GetAccessFlagsEx |
+                                                          SMF_AuthorizeFileScope_Named | SMF_AuthorizeWorkUnitScope_Named;
+    static const SecFeatureSet s_safeFeatureMask = s_implementedFeatureMask | SMF_CreateSettingMap;
+
+    MapStringTo<CUserAccess*> userAccessMap;
+
+    void readUserAccess(IPropertyTree* secMgrCfg, IPropertyTree* bindConfig)
+    {
+        StringBuffer s;
+        Owned<IPropertyTreeIterator> iter = secMgrCfg->getElements("useraccess");
+        ForEach(*iter)
+        {
+            IPropertyTree& userSettings = iter->query();
+
+            StringBuffer userName, password;
+            if (!readUserNamePassword(userSettings, userName, password))
+                continue;
```

Comment:
It's up to you, a bad config can confuse things especially while using this for something like a regression test.. so stopping ESP can be good to prevent confusion.

Reply:
Changing to throw exception